File: src/web/src/types/user-specimen.ts
// UserSpecimen types
export interface UserSpecimenDto {
  userSpecimenId: string;
  userId: string;
  commonName: string;
  scientificName: string | null;
  alias: string | null;
  rockFamily: string | null;
  species: string | null;
  variety: string | null;
  materialType: string;
  mohsHardnessMin: number | null;
  mohsHardnessMax: number | null;
  tumblingDifficulty: string | null;
  recommendedGritSequence: string | null;
  specialConsiderations: string | null;
  notes: string | null;
  isPublic: boolean;
  basedOnSpecimenId: string | null;
  dateCreated: string;
  aiConfidenceScore: number | null;
  aiIsKnownSpecimen: boolean | null;
}

export interface UserSpecimenListDto {
  userSpecimenId: string;
  commonName: string;
  scientificName: string | null;
  materialType: string;
  tumblingDifficulty: string | null;
  isPublic: boolean;
  dateCreated: string;
  aiConfidenceScore: number | null;
  aiIsKnownSpecimen: boolean | null;
}

export interface CreateUserSpecimenRequest {
  commonName: string;
  scientificName?: string | null;
  alias?: string | null;
  rockFamily?: string | null;
  species?: string | null;
  variety?: string | null;
  materialType?: string;
  mohsHardnessMin?: number | null;
  mohsHardnessMax?: number | null;
  tumblingDifficulty?: string | null;
  recommendedGritSequence?: string | null;
  specialConsiderations?: string | null;
  notes?: string | null;
  isPublic?: boolean;
  basedOnSpecimenId?: string | null;
  aiConfidenceScore?: number | null;
  aiIsKnownSpecimen?: boolean | null;
}

export interface UpdateUserSpecimenRequest {
  commonName: string;
  scientificName?: string | null;
  alias?: string | null;
  rockFamily?: string | null;
  species?: string | null;
  variety?: string | null;
  materialType?: string;
  mohsHardnessMin?: number | null;
  mohsHardnessMax?: number | null;
  tumblingDifficulty?: string | null;
  recommendedGritSequence?: string | null;
  specialConsiderations?: string | null;
  notes?: string | null;
  isPublic?: boolean;
  aiConfidenceScore?: number | null;
  aiIsKnownSpecimen?: boolean | null;
}

// Combined DTO for UI dropdowns/pickers - includes both system and user specimens
export interface SpecimenOptionDto {
  id: string;
  commonName: string;
  scientificName: string | null;
  alias: string | null;
  materialType: string;
  tumblingDifficulty: string | null;
  mohsHardnessMax: number | null;  // For hardness warning calculation
  source: 'system' | 'user';
  isOwned: boolean;
  basedOnSpecimenId: string | null;
}

// Filters for user specimens list
export interface UserSpecimenFilters {
  search?: string;
  materialType?: string;
  sortBy?: 'commonName' | 'dateCreated';
  sortOrder?: 'asc' | 'desc';
}

// AI Specimen Lookup types
export interface SpecimenLookupRequest {
  commonName: string;
  sourceUrl?: string | null;
  sourceName?: string | null;
  sourceDescription?: string | null;
}

export interface SpecimenLookupResponse {
  success: boolean;
  error: string | null;
  data: SpecimenLookupData | null;
}

export interface SpecimenLookupData {
  commonName: string;
  scientificName: string | null;
  alias: string | null;
  rockFamily: string | null;
  species: string | null;
  variety: string | null;
  materialType: string;
  mohsHardnessMin: number | null;
  mohsHardnessMax: number | null;
  tumblingDifficulty: string | null;
  recommendedGritSequence: string | null;
  specialConsiderations: string | null;
  isKnownSpecimen: boolean;
  confidenceScore: number;
  confidenceReason: string | null;
}

File: src/api/MyUglyRocks.Core/Entities/UserSpecimen.cs
namespace MyUglyRocks.Core.Entities;

public class UserSpecimen : SoftDeletableEntity
{
    public Guid UserSpecimenId { get; set; }
    public Guid UserId { get; set; }
    public required string CommonName { get; set; }
    public string? ScientificName { get; set; }
    public string? Alias { get; set; }
    public string? RockFamily { get; set; }
    public string? Species { get; set; }
    public string? Variety { get; set; }
    public SpecimenMaterialType MaterialType { get; set; } = SpecimenMaterialType.Rock;
    public decimal? MohsHardnessMin { get; set; }
    public decimal? MohsHardnessMax { get; set; }
    public TumblingDifficulty? TumblingDifficulty { get; set; }
    public string? RecommendedGritSequence { get; set; }
    public string? SpecialConsiderations { get; set; }
    public string? Notes { get; set; }
    public bool IsPublic { get; set; }
    public Guid? BasedOnSpecimenId { get; set; }

    // Navigation properties
    public virtual User User { get; set; } = null!;
    public virtual Specimen? BasedOnSpecimen { get; set; }
    public virtual ICollection<CycleSpecimen> CycleSpecimens { get; set; } = [];
    public virtual ICollection<InventorySpecimen> InventorySpecimens { get; set; } = [];
}
